test_bucket_delete_with_objects fails in about 50% of runs on 4.21 - investigate and fix if needed

test_bucket_delete_with_objects fails in about 50% of runs on 4.21 - investigate and fix if needed

Relevant runs: 

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/43519/1998957/1998999/log

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/44528/2028539/2028577/log

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/44903/2039562/2039603/log

https://reportportal-ocs4.apps.dno.ocp-hub.prod.psi.redhat.com/ui/#ocs/launches/994/45028/2041536/2041573/log

2 types of failures:

1)    raise error_class(parsed_response, operation_name)
E           botocore.exceptions.ClientError: An error occurred (504) when calling the CreateBucket operation (reached max retries: 8): Gateway Timeout



venv/lib64/python3.11/site-packages/botocore/client.py:1058: ClientError

2)           raise CommandFailed(
                f"Error during execution of command: {masked_cmd}."
                f"\nError is {masked_stderr}"
            )
E               ocs_ci.ocs.exceptions.CommandFailed: Error during execution of command: oc --kubeconfig /home/jenkins/current-cluster-dir/openshift-cluster-dir/auth/kubeconfig -n s3cli-spc4ori5tmdhw rsh s3cli-0 sh -c 'AWS_CA_BUNDLE=/tmp/ca-bundle.crt AWS_ACCESS_KEY_ID=* AWS_SECRET_ACCESS_KEY=* AWS_DEFAULT_REGION=us-east-2 aws s3 --endpoint=*****  sync /test_objects/ s3://s3-bucket-2e90e26a54fd4478887b99c5036925'.
E               Error is upload failed: ../test_objects/rome.jpg to s3://s3-bucket-2e90e26a54fd4478887b99c5036925/rome.jpg An error occurred (InvalidBucketState) when calling the UploadPart operation: The request is not valid with the current state of the bucket.
E               upload failed: ../test_objects/enwik8 to s3://s3-bucket-2e90e26a54fd4478887b99c5036925/enwik8 An error occurred (InvalidBucketState) when calling the UploadPart operation: The request is not valid with the current state of the bucket.
E               upload failed: ../test_objects/steve.webm to s3://s3-bucket-2e90e26a54fd4478887b99c5036925/steve.webm An error occurred (InvalidBucketState) when calling the UploadPart operation: The request is not valid with the current state of the bucket.
E               command terminated with exit code 1



ocs_ci/utility/utils.py:878: CommandFailed

--- a/tests/functional/object/mcg/test_bucket_deletion.py
+++ b/tests/functional/object/mcg/test_bucket_deletion.py
@@ -27,6 +27,7 @@ from ocs_ci.ocs.constants import AWSCLI_TEST_OBJ_DIR
 from ocs_ci.ocs.exceptions import CommandFailed
 from ocs_ci.ocs.ocp import OCP
 from ocs_ci.ocs.resources.objectbucket import MCGS3Bucket
+from ocs_ci.utility.retry import retry
 
 logger = logging.getLogger(__name__)
 ERRATIC_TIMEOUTS_SKIP_REASON = "Skipped because of erratic timeouts"
@@ -123,11 +124,29 @@ class TestBucketDeletion(MCGTest):
         Negative test with deletion of bucket has objects stored in.
 
         """
-        bucketname = bucket_factory(bucketclass=bucketclass_dict)[0].name
+        logger.info(
+            f"Creating bucket with interface={interface}, "
+            f"bucketclass_dict={bucketclass_dict}"
+        )
+        bucket = bucket_factory(bucketclass=bucketclass_dict)[0]
+        bucketname = bucket.name
+        logger.info(f"Bucket {bucketname} created successfully")
 
         data_dir = AWSCLI_TEST_OBJ_DIR
         full_object_path = f"s3://{bucketname}"
-        sync_object_directory(awscli_pod_session, data_dir, full_object_path, mcg_obj)
+        logger.info(
+            f"Syncing objects from {data_dir} to {full_object_path} "
+            f"(endpoint: {mcg_obj.s3_internal_endpoint})"
+        )
+        retry(
+            (CommandFailed, botocore.exceptions.ClientError),
+            tries=3,
+            delay=30,
+            backoff=2,
+        )(sync_object_directory)(
+            awscli_pod_session, data_dir, full_object_path, mcg_obj
+        )
+        logger.info(f"Objects synced successfully to {bucketname}")
 
         logger.info(f"Deleting bucket: {bucketname}")
         if interface == "S3":

